### PR TITLE
Refactor remove reducers from filters

### PR DESCRIPTION
**Pull request by joonatank**
## 🛠️ Changelog
[//]: # "Describe the changes in this pull request here."

- refactor: rest of admin filters to use query param based system instead of reducers.
- fix: applied filters are copyable (by copying a link).
- fix: filters persist between page changes (ex. clicking on a list item and pressing back on the browser).
- add: auto fetch to query filters. If there are more than 100 units or reservation units then all of them are available for selecting.

## 🧪 Test plan
[//]: # "Help your fellow reviewer and write a short description of what's the fastest way to test your changes."

- Manual testing: filters on admin pages (application-round / allocation not touched) should work as before / correctly. Ex. reservations / reservation-units.

## 🎫 Tickets
[//]: # "This pull request resolves all or part of the following ticket(s)"

- [TILA-3419](https://helsinkisolutionoffice.atlassian.net/browse/TILA-3419)

[TILA-3419]: https://helsinkisolutionoffice.atlassian.net/browse/TILA-3419?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/admin-ui/src/component/filters/ReservationUnitFilter.tsx [hunk 1]
```diff
@@ -1,59 +1,28 @@
...
+    onCompleted: async (d) => {
+      if (d?.reservationUnits?.pageInfo.hasNextPage) {
+        await fetchMore({
+          variables: {
+            after: d.reservationUnits.pageInfo.endCursor,
+          },
+        });
+      }
+    },
   });
```
Comment (joonatank) [inline apps/admin-ui/src/component/filters/ReservationUnitFilter.tsx:16]: FIX: This doesn't work for more than 200 objects. `onComplete` is only ever called once. Have to use `useEffect` for this.

File: apps/admin-ui/src/component/filters/UnitFilter.tsx [hunk 2]
```diff
@@ -1,58 +1,48 @@
...
+    // auto fetch more (there is no limit, expect number of them would be a few hundred, but in theory this might cause problems)
+    onCompleted: async (d) => {
+      if (d?.units?.pageInfo.hasNextPage) {
+        await fetchMore({
+          variables: {
+            after: d.units.pageInfo.endCursor,
+          },
+        });
+      }
+    },
```
Comment (joonatank) [inline apps/admin-ui/src/component/filters/UnitFilter.tsx:37]: FIX: some issue as with other `onComplete`

File: apps/admin-ui/src/component/reservations/Filters.tsx [hunk 3]
```diff
@@ -52,192 +31,138 @@ const MoreWrapper = styled(ShowAllContainer)`
...
-  reservationUnit: [],
-  paymentStatuses: [],
-  reservationState: [],
-  textSearch: "",
-  begin: "",
-  end: "",
-  minPrice: "",
-  maxPrice: "",
-};
+function DateInputFilter({ name }: { name: string }) {
```
Comment (joonatank) [inline apps/admin-ui/src/component/reservations/Filters.tsx:34]: Why isn't this with the other filter components?

File: apps/admin-ui/src/component/reservations/Filters.tsx [hunk 4]
```diff
@@ -52,192 +31,138 @@ const MoreWrapper = styled(ShowAllContainer)`
...
+    } else {
+      setParams(params, { replace: true });
+    }
+  };
 
-  const [state, dispatch] = useReducer(
-    getReducer<FilterArguments>(initialEmptyState),
-    initialEmptyState
+  const label = t(`filters.label.${name}`);
+  // TODO make the translation empty. no placeholder on purpose
```
Comment (joonatank) [inline apps/admin-ui/src/component/reservations/Filters.tsx:51]: ??

File: apps/admin-ui/src/component/reservations/Filters.tsx [hunk 5]
```diff
@@ -52,192 +31,138 @@ const MoreWrapper = styled(ShowAllContainer)`
...
+  const placeholder = t(`filters.placeholder.${name}`);
+  return (
+    <DateInput
+      language="fi"
+      // TODO ids should be unique and the name nor the tr key is not
```
Comment (joonatank) [inline apps/admin-ui/src/component/reservations/Filters.tsx:56]: name isnt?

File: apps/admin-ui/src/component/reservations/ReservationsDataLoader.tsx [hunk 6]
```diff
@@ -1,70 +1,104 @@
...
-import { filterNonNullable } from "common/src/helpers";
-
-type Props = {
-  filters: FilterArguments;
-  defaultFiltering: QueryReservationsArgs;
-};
+import { filterNonNullable, toNumber } from "common/src/helpers";
+import { useSearchParams } from "react-router-dom";
+
+function transformPaymentStatusSafe(t: string): OrderStatus | null {
```
Comment (joonatank) [inline apps/admin-ui/src/component/reservations/ReservationsDataLoader.tsx:19]: could move to `helpers` etc. where other enum conversions are

File: apps/admin-ui/src/component/reservations/ReservationsDataLoader.tsx [hunk 7]
```diff
@@ -1,70 +1,104 @@
...
+  }
+}
 
-function mapFilterParams(
-  params: FilterArguments,
-  defaultParams: QueryReservationsArgs
-): QueryReservationsArgs {
-  const emptySearch =
-    values(params).filter((v) => !(v === "" || v.length === 0)).length === 0;
+function transformStateSafe(t: string): State | null {
```
Comment (joonatank) [inline apps/admin-ui/src/component/reservations/ReservationsDataLoader.tsx:38]: move??

File: apps/admin-ui/src/hooks/useUnitOptions.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import { useEffect } from "react";
+import { gql } from "@apollo/client";
+import { filterNonNullable } from "common/src/helpers";
+import { useUnitsFilterQuery } from "@gql/gql-types";
+
+// exporting so it doesn't get removed
+// TODO combine with other options queries so we only make a single request for all of them
```
Comment (joonatank) [inline apps/admin-ui/src/hooks/useUnitOptions.ts:7]: remove comment

File: packages/common/src/helpers.ts [hunk 9]
```diff
@@ -7,6 +7,23 @@ export function filterNonNullable<T>(
   return arr?.filter((n): n is NonNullable<T> => n != null) ?? [];
 }
 
+/// Safe string -> number conversion
+/// handles the special cases of empty string and NaN with type safety
+export function toNumber(filter: string | null): number | null {
```
Comment (joonatank) [inline packages/common/src/helpers.ts:12]: pk's are also always > 0, should we filter those out?

**Review by vincit-matu — APPROVED** [review]
LGTM 👍

**Pull request merged**